<!-- # --- -->
title: "Untitled"
author: "vitoria"
date: "2025-04-16"
output: html_document
---

```{r message=FALSE, warning=FALSE, include=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

# Pacotes do CRAN necessários
cran_pkgs <- c(
  "data.table", "dplyr", "future", "future.apply", "ggplot2", "rlang", "irlba", 
  "tidyr", "patchwork", "stringi", "fastmatch", "RcppRoll", "scales", "Rcpp", 
  "tidyselect", "vctrs", "lifecycle", "dotCall64", "RcppEigen", "png", "reticulate"
)

# Pacotes do Bioconductor necessários
bioc_pkgs <- c(
  "GenomeInfoDb", "GenomicRanges", "IRanges", "Rsamtools", "S4Vectors", 
  "SeuratObject", "Biostrings", "XVector", "Rhtslib", "BiocParallel", 
  "UCSC.utils", "WGCNA"
)

# Função pra instalar se faltar
install_missing <- function(pkgs, installer) {
  for (pkg in pkgs) {
    if (!requireNamespace(pkg, quietly = TRUE)) {
      message(paste("Instalando:", pkg))
      installer(pkg)
    } else {
      message(paste("Já está instalado:", pkg))
    }
  }
}

# Instala os pacotes
install_missing(cran_pkgs, install.packages)
install_missing(bioc_pkgs, BiocManager::install)

if (!requireNamespace("remotes", quietly = TRUE))
    install.packages("remotes")
remotes::install_github("stuart-lab/signac", ref = "develop")

devtools::install_github("immunogenomics/presto")
```

```{r message=FALSE, warning=FALSE}
# Load necessary libraries
library(dplyr)
library(Seurat)
library(Signac)
library(GenomeInfoDb)
library(ggplot2)
library(patchwork)
library(Matrix)
library(GENIE3)
# library(variancePartition)
# library(BiocParallel)
# library(WGCNA)
```

# Functions

```{r}
#' Builds a Seurat object from snRNA-seq data
#'
#' @param matrix_file String - Path to a .mtx.gz file
#' @param genes_file String - Path to file with gene names
#' @param cells_file String - Path to file with cell names
#' @param project_name String - Name of your project/dataset
#' @param min.cells Integer - Minimum number of cells a gene must be detected in to be retained
#' @param min.features Integer - Minimum number of genes a cell must express to be included
#' @param nfeatures Integer - Number of top variable features to select.
#' @return seurat_obj - SeuratObject for downstream analyses
process_seurat_object <- function(matrix_file, genes_file, cells_file, group_file,
                                  project_name = "SeuratProject",
                                  min_cells = 3, min_features = 100) {
  
  # Load expression matrix
  expr_matrix <- readMM(matrix_file)
  
  # Load gene and cell names
  genes <- read.csv(genes_file, header = TRUE)
  cells <- read.csv(cells_file, header = TRUE)
  
  # Define row and column names
  rownames(expr_matrix) <- make.unique(genes$x)
  colnames(expr_matrix) <- cells$x
  
  # Create object
  seurat_obj <- CreateSeuratObject(counts = expr_matrix, 
                                   project = project_name, 
                                   min.cells = min_cells, 
                                   min.features = min_features)
  
  return(seurat_obj)
}

#' Builds a Seurat object from snRNA-seq data
#'
#' @param seurat_obj SeuratObject
#' @param nfeatures Integer - Number of features to filter if your dataset is large (e.g. top 10000 genes)
#' @param group_by String - Name of the variable you want to group your data by (condition, treatment, etc)
#' @param ident_1 String - Name of the first group considering the condition above
#' @param ident_2 String - Name of the second group considering the condition above
#' @param method String - Statistical test for differential expression ("wilcox", "t", "bimod", "LR")
#' @param logfc_threshold Integer - Threshold for FoldChange filtering (e.g. > 2)
#' @param pval_adj_threshold Float - Threshold for adjusted p value (e.g. <= 0.01) 
#' @param save Boolean - If you want to save your DEGs as a CSV
#' @param output_path String - Path to save the CSVs
#' @param prefix String - Prefix for the CSV filenames
#' @return list with p-adjusted DEGs and filtered DEGs
run_DE_analysis <- function(seurat_obj,
                            nfeatures = 20000,
                            group_by = "condition",
                            ident_1 = "MDD",
                            ident_2 = "Control",
                            method = "wilcox",
                            logfc_threshold = 1,
                            pval_adj_threshold = 0.05,
                            save = FALSE,
                            output_path = "",
                            prefix = "") {
  
  # Find DEGs
  degs <- FindMarkers(seurat_obj,
                      ident.1 = ident_1,
                      ident.2 = ident_2,
                      group.by = group_by,
                      test.use = method)
  
  # Adjust p value
  degs$p_val_adj <- p.adjust(degs$p_val, method = "fdr")
  
  # Filter by p value and FC
  degs_filtered <- degs %>%
    filter(p_val_adj < pval_adj_threshold, abs(avg_log2FC) > logfc_threshold)

  # Save if requested
  if (save){
    write.csv(degs,
              file = file.path(output_path, paste0(prefix, "_results_adj.csv")),
              row.names = TRUE)
  
    write.csv(degs_filtered,
              file = file.path(output_path, paste0(prefix, "_results_filtered.csv")),
              row.names = TRUE)    
  }  

  return(list(degs_all = degs, degs_filtered = degs_filtered))
}

```

# GSE213982

## Load data

```{r}
gse_path <- "data/snrnaseq/GSE213982/"

# Build the Seurat object
GSE213982 <- process_seurat_object(
  matrix_file = paste0(gse_path, "GSE213982_combined_counts_matrix.mtx.gz"),
  genes_file = paste0(gse_path,  "GSE213982_combined_counts_matrix_genes_rows.csv.gz"),
  cells_file = paste0(gse_path, "GSE213982_combined_counts_matrix_cells_columns.csv.gz"),
  project_name = "GSE213982_snRNA"
)

# Gather meta data
cell_info <- do.call(rbind, strsplit(colnames(GSE213982), "\\."))
GSE213982$sample_id <- cell_info[,1]
GSE213982$barcode <- cell_info[,2]
GSE213982$cell_type <- cell_info[,3]
GSE213982$cell_subtype <- cell_info[,4]

# Gather groups information
condition_info <- read.csv(paste0(gse_path, "/groups.csv"))
GSE213982$condition <- condition_info$Condition[match(GSE213982$sample_id,
                                                       condition_info$Sample)]
GSE213982$condition[GSE213982$condition == "Case"] <- "MDD"

# Normalization
GSE213982 <- NormalizeData(GSE213982, normalization.method = "LogNormalize")

save(GSE213982, file=paste0(gse_path, "GSE213982_counts.RData"))
```
## Analysis without considering cell types

```{r}
load(paste0(gse_path, "GSE213982_counts.RData"))

# Select Variable Features to reduce the dataset size
# GSE213982 <- FindVariableFeatures(seurat_obj,
#                                    selection.method = "vst",
#                                    nfeatures = 15000)

results_GSE213982 <- run_DE_analysis(GSE213982, save = FALSE)
GSE213982 <- NULL
gc()
```

## Analysis considering cell types

```{r message=FALSE, warning=FALSE, include=FALSE}
gse_path <- "data/snrnaseq/GSE213982/"
load(paste0(gse_path, "GSE213982_counts.RData"))

GSE213982 <- FindVariableFeatures(GSE213982,
                                     selection.method = "vst",
                                     nfeatures = 15000)

GSE213982_markers <- list()
cell_types <- setdiff(unique(GSE213982@meta.data$cell_type), "Mix")

  
for (ct in cell_types) {
  seurat_sub <- subset(GSE213982,
                       subset = cell_type == ct)
  # Select Variable Features to reduce the dataset size

  seurat_sub <- NormalizeData(seurat_sub,
                              normalization.method = "LogNormalize")
  Idents(seurat_sub) <- "condition"

  markers <- FindMarkers(seurat_sub, ident.1 = "MDD",
                         ident.2 = "Control",
                         test.use = "wilcox")
  
  GSE213982_markers[[ct]] <- markers
  gc()
}

# filter by p and FC
GSE213982_markers_filtered <- list()
for (ct in cell_types) {
  markers <- GSE213982_markers[[ct]]
  
  markers_filtered <- markers %>%
    filter(p_val_adj < 0.05, abs(avg_log2FC) >= 1)

  GSE213982_markers_filtered[[ct]] <- markers_filtered
}

# save(GSE213982_markers_filtered,
#      file=paste0(gse_path, "divided_by_type/GSE213982_markers_type.RData"))
```


## GSE144136

```{r}
gse_path <- "data/snrnaseq/GSE144136/"

GSE144136 <- process_seurat_object(
  matrix_file = paste0(gse_path, "GSE144136_GeneBarcodeMatrix_Annotated.mtx.gz"),
  genes_file = paste0(gse_path, "GSE144136_GeneNames.csv.gz"),
  cells_file = paste0(gse_path, "GSE144136_CellNames.csv.gz"),
  project_name = "GSE144136_snRNA"
)

cell_info <- do.call(rbind, strsplit(colnames(GSE144136), "\\_"))
GSE144136$cell_type <- cell_info[,1]

GSE144136@meta.data$condition <- ifelse(grepl("Control", rownames(GSE144136@meta.data)), 
                                    "Control", 
                                    ifelse(grepl("Suicide", rownames(GSE144136@meta.data)), 
                                           "Suicide", 
                                           NA))

GSE144136$condition[GSE144136$condition == "Suicide"] <- "MDD"

GSE144136@meta.data$cell_type <- dplyr::case_when(
  GSE144136@meta.data$cell_type == "Astros" ~ "Ast",
  GSE144136@meta.data$cell_type == "Inhib" ~ "InN",
  GSE144136@meta.data$cell_type == "Ex" ~ "ExN",
  GSE144136@meta.data$cell_type == "Oligos" ~ "Oli",
  GSE144136@meta.data$cell_type == "OPCs" ~ "OPC",
  grepl("^Micro/Macro", GSE144136@meta.data$cell_type) ~ "Mic",
  grepl("^Endo", GSE144136@meta.data$cell_type) ~ "End",

  TRUE ~ GSE144136@meta.data$cell_type  
)

GSE144136 <- NormalizeData(GSE144136, normalization.method = "LogNormalize")

save(GSE144136, file=paste0(gse_path, "GSE144136_counts.RData"))
```

```{r message=FALSE, warning=FALSE}
load(paste0(gse_path, "/GSE144136_counts.RData"))

results_GSE144136 <- run_DE_analysis(GSE144136, save = FALSE)
GSE144136 <- NULL
gc()
```

## Analysis considering cell types


```{r eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}
gse_path <- "data/snrnaseq/GSE144136/"
load(paste0(gse_path, "/GSE144136_counts.RData"))

GSE144136 <- FindVariableFeatures(GSE144136,
                                     selection.method = "vst",
                                     nfeatures = 1000)

GSE144136_markers <- list()
cell_types <- setdiff(unique(GSE144136@meta.data$cell_type), "Mix")

for (ct in cell_types) {
  seurat_sub <- subset(GSE144136, subset = cell_type == ct)
  seurat_sub <- NormalizeData(seurat_sub, normalization.method = "LogNormalize")

  Idents(seurat_sub) <- "condition"

  markers <- FindMarkers(seurat_sub, ident.1 = "MDD", ident.2 = "Control",
                         test.use = "wilcox")

  GSE144136_markers[[ct]] <- markers
}

# filter by p and FC
GSE144136_markers_filtered <- list()
for (ct in cell_types) {
  markers <- GSE144136_markers[[ct]]
  
  markers_filtered <- markers %>%
    filter(p_val_adj < 0.05, abs(avg_log2FC) >= 1)

  GSE144136_markers_filtered[[ct]] <- markers_filtered
}

# save(GSE144136_markers_filtered,
#      file=paste0(gse_path, "divided_by_type/GSE144136_markers_type.RData"))
```

## Getting the intersection between snRNA-seq and target prediction

```{r}
# Load target_intersection (303 targets)
load(file='data/target_prediction/targets_intersection.RData')

# GSE213982
degs_GSE213982 <- rownames(results_GSE213982$degs_filtered)
degs_GSE144136 <- rownames(results_GSE144136$degs_filtered)

# Intersection between all of them
snrnaseq_intersection <- intersect(degs_GSE213982, degs_GSE144136)
all_intersections <- intersect(snrnaseq_intersection, targets_intersection)
```

# Comparing cell types

### Type by type

```{r message=FALSE, warning=FALSE, include=FALSE}
gene <- "NR4A1"
nr4a1_expressions <- list()
cell_types <- setdiff(unique(GSE213982@meta.data$cell_type), "Mix")
for (ct in cell_types) {
  
  gene_data <- FetchData(GSE213982, vars = c("NR4A1", "condition", "cell_type"))
  gene_data_filtered <- gene_data[gene_data$cell_type == ct, ]

  # Create violin plot
  nr4a1_condition_exp <- ggplot(gene_data_filtered, aes(x = condition,
                                                     y = NR4A1)) +
                          geom_boxplot(alpha = 0.7, outlier.alpha = 0.5) +
                          geom_jitter(width = 0.2, alpha = 0.3, size = 0.5) +
                          # geom_violin(alpha = 0.7, trim = FALSE) +
                          # geom_boxplot(width = 0.1, fill = "white",
                          #              alpha = 0.8) +
                          theme_light() +
                          labs(title = paste0(gene, " expression - ", ct),
                               x = "Condition", 
                               y = "NR4A1 Expression") +
                          theme(legend.position = "none",
                                plot.title = element_text(hjust = 0.5,
                                                          size = 14,
                                                          face = "bold"))

  ggsave(paste0(gse_path, gene, "_", ct, ".png"), nr4a1_condition_exp)
}
```

## Cell type

```{r message=FALSE, warning=FALSE}
# Load target_intersection (303 targets)
load(file='data/target_prediction/targets_intersection.RData')

intersection_celltypes <- list()
for (ct in cell_types) {
  intersection_celltypes[[ct]] <- intersect(rownames(GSE144136_markers_filtered[[ct]]),
                                            rownames(GSE213982_markers_filtered[[ct]]))
}

intersection_with_targets <- list()
for (ct in cell_types) {
  intersection_with_targets[[ct]] <- intersect(targets_intersection,
                                               intersection_celltypes[[ct]])
}

save(intersection_celltypes, intersection_with_targets, file = "intersections.RData")
```

# WGCNA coexpression

```{r message=FALSE, warning=FALSE}
gse_path <- "data/snrnaseq/GSE213982/"
load(paste0(gse_path, "GSE213982_variablefeatures.RData"))

traitData <- GSE213982_features@meta.data[, c("condition", "cell_type")]

GSE213982_matrix <- GetAssayData(GSE213982_features, assay = "RNA", slot = "data")
variable_genes <- VariableFeatures(GSE213982_features)
GSE213982_matrix <- GSE213982_matrix[variable_genes, ]
GSE213982_matrix <- t(as.matrix(GSE213982_matrix))  

GSE213982_features <- NULL
gc()

GSE213982_gsg <- goodSamplesGenes(GSE213982_matrix, verbose = 3)
if (!GSE213982_gsg$allOK) {
  GSE213982_matrix <- GSE213982_matrix[GSE213982_gsg$goodSamples, GSE213982_gsg$goodGenes]
}

GSE213982_gsg <- NULL
```

```{r message=FALSE, warning=FALSE}
softPower <- 5  # replace with chosen power
adjacency <- adjacency(GSE213982_matrix, power = softPower)

TOM <- TOMsimilarity(adjacency)
dissTOM <- 1 - TOM

geneTree <- hclust(as.dist(dissTOM), method = "average")
dynamicMods <- cutreeDynamic(dendro = geneTree, distM = dissTOM,
                             deepSplit = 3, pamRespectsDendro = FALSE,
                             minClusterSize = 20)
moduleColors <- labels2colors(dynamicMods)
```

```{r message=FALSE, warning=FALSE}
MEs <- moduleEigengenes(GSE213982_matrix, colors = moduleColors)$eigengenes
dissimilarity <- 1 - cor(MEs)
merge <- mergeCloseModules(GSE213982_matrix, moduleColors, cutHeight = 0.25, verbose = 3)
moduleColors <- merge$colors
MEs <- merge$newMEs

traitData$condition <- as.numeric(factor(traitData$condition))
traitData$cell_type <- as.numeric(factor(traitData$cell_type))

moduleTraitCor <- cor(MEs, traitData, use = "p")
moduleTraitPvalue <- corPvalueStudent(moduleTraitCor, nrow(GSE213982_matrix))

textMatrix <- paste(signif(moduleTraitCor, 2), "\n(",
                    signif(moduleTraitPvalue, 1), ")", sep = "")
dim(textMatrix) <- dim(moduleTraitCor)

labeledHeatmap(Matrix = moduleTraitCor,
               xLabels = colnames(traitData),
               yLabels = names(MEs),
               ySymbols = names(MEs),
               colorLabels = FALSE,
               colors = blueWhiteRed(50),
               textMatrix = textMatrix,
               setStdMargins = FALSE,
               cex.text = 0.5,
               zlim = c(-1,1),
               main = "Module-trait relationships (merged)")

```
```{r message=FALSE, warning=FALSE}
# Get genes in the modules
names(moduleColors) <- colnames(GSE213982_matrix)
module_genes <- list()
for (color in unique(moduleColors)){
  module_genes[[color]] <- names(moduleColors[moduleColors == color])
}

all_intersections <- list()
for (color in names(module_genes)){
  all_intersections[[color]] <- intersect(targets_intersection, module_genes$color)
}

load(file='data/target_prediction/targets_intersection.RData')

```
```{r message=FALSE, warning=FALSE}

top10_genes <- list()
for (color in names(MEs)) {
  kME <- cor(GSE213982_matrix, MEs[, color])  
  kME_vector <- as.vector(kME)
  names(kME_vector) <- rownames(kME)  
  kME_sorted <- sort(abs(kME_vector), decreasing = TRUE)
  top_genes <- names(kME_sorted)[1:10] # gene names

  top10_genes[[color]] <- data.frame(
    gene = top_genes,
    kME = kME_vector[top_genes]  # signed correlations
  )
}

```
```{r message=FALSE, warning=FALSE}

# moduleTraitCor is a matrix: rows = modules, cols = traits
# moduleTraitPvalue is the corresponding p-value matrix

# Thresholds
cor_threshold <- 0.4
pval_threshold <- 0.05

# Find modules with |correlation| > 0.4 and p-value < 0.05 for each trait
significant_modules <- list()

for(trait in colnames(moduleTraitCor)) {
  sig_mods <- rownames(moduleTraitCor)[
    abs(moduleTraitCor[, trait]) > cor_threshold &
    moduleTraitPvalue[, trait] < pval_threshold
  ]
  significant_modules[[trait]] <- sig_mods
}

significant_modules
```

## 14

```{r message=FALSE, warning=FALSE}
# gse_path <- "data/snrnaseq/GSE213982/"
# load(paste0(gse_path, "GSE213982_variablefeatures.RData"))

traitData <- GSE144136@meta.data[, c("condition", "cell_type")]

GSE144136_matrix <- GetAssayData(GSE144136, assay = "RNA", slot = "data")
variable_genes <- VariableFeatures(GSE144136)
GSE144136_matrix <- GSE144136_matrix[variable_genes, ]
GSE144136_matrix <- t(as.matrix(GSE144136_matrix))  

GSE213982_features <- NULL
gc()

GSE144136_gsg <- goodSamplesGenes(GSE144136_matrix, verbose = 3)
if (!GSE144136_gsg$allOK) {
  GSE144136_matrix <- GSE144136_matrix[GSE144136_gsg$goodSamples, GSE144136_gsg$goodGenes]
}

GSE144136_gsg <- NULL
```

```{r message=FALSE, warning=FALSE}
softPower <- 5  # replace with chosen power
adjacency <- adjacency(GSE144136_matrix, power = softPower)

TOM <- TOMsimilarity(adjacency)
dissTOM <- 1 - TOM

geneTree <- hclust(as.dist(dissTOM), method = "average")
dynamicMods <- cutreeDynamic(dendro = geneTree, distM = dissTOM,
                             deepSplit = 3, pamRespectsDendro = FALSE,
                             minClusterSize = 20)
moduleColors <- labels2colors(dynamicMods)
```

# variance algo

```{r message=FALSE, warning=FALSE}
library(BiocParallel)
library(variancePartition)

# Prepare your data
# Assume `seurat_obj` is your Seurat object with normalized data and metadata
GSE144136_matrix <- GetAssayData(GSE144136, assay = "RNA", slot = "data")

meta_data <- GSE144136@meta.data

meta_data$condition <- as.numeric(factor(meta_data$condition))
meta_data$cell_type <- as.numeric(factor(meta_data$cell_type))

GSE144136 <- NULL
gc()

# Create a formula describing the variables to partition
form <- ~ (1|orig.ident) + condition + cell_type

param <- SnowParam(workers = 4, type = "SOCK", progressbar = FALSE)

# Define chunk size
chunk_size <- 1000
gene_names <- rownames(GSE144136_matrix)
n_genes <- length(gene_names)

# Create empty list to store results
varPart_list <- list()
gc()

# Loop over chunks
for (i in seq(1, n_genes, by = chunk_size)) {
  cat("Processing genes", i, "to", min(i + chunk_size - 1, n_genes), "\n")
  
  # Subset matrix
  gene_subset <- gene_names[i:min(i + chunk_size - 1, n_genes)]
  matrix_chunk <- GSE144136_matrix[gene_subset, ]

  # Run model
  varPart_chunk <- fitExtractVarPartModel(matrix_chunk, form, meta_data, 
                                          BPPARAM = SnowParam(4))  # Adjust cores if needed
  
  # Store result
  varPart_list[[length(varPart_list) + 1]] <- varPart_chunk
  
  # Clean up
  rm(matrix_chunk)
  gc()
}

# Combine results
varPart_combined <- do.call(rbind, varPart_list)
varPart_combined$gene <- rownames(varPart_combined)

top_genes <- head(varPart_combined[order(varPart_combined$condition, decreasing = TRUE), c("condition","gene")], 100)

load(file='data/target_prediction/targets_intersection.RData')
intersect(targets_intersection, top_genes$gene)

save(top_genes, file=paste0(gse_path, "GSE144136_variance.RData"))
```
# network algo


```{r message=FALSE, warning=FALSE}
library(GENIE3)

# Expressão como matriz
GSE144136_matrix <- GetAssayData(GSE144136, assay = "RNA", slot = "data")
expr_matrix <- as.matrix(GSE144136_matrix)

# Define parâmetros
chunk_size <- 100  # ajuste conforme sua RAM
gene_names <- rownames(expr_matrix)
n_genes <- length(gene_names)
regulators <- gene_names  

# Lista para armazenar os resultados
weight_list <- list()

# Loop em chunks de targets
for (i in seq(1, n_genes, by = chunk_size)) {
  cat("Processing targets", i, "to", min(i + chunk_size - 1, n_genes), "\n")
  
  target_genes <- gene_names[i:min(i + chunk_size - 1, n_genes)]
  
  # Rodar GENIE3 só para os targets especificados
  weights_chunk <- GENIE3(expr_matrix,
                          regulators = regulators,
                          targets = target_genes,
                          nTrees = 500,
                          nCores = 1))  
  
  weight_list[[length(weight_list) + 1]] <- weights_chunk
  
  rm(weights_chunk)
  gc()
}

# Combinar matrizes
weight_matrix_combined <- Reduce("+", weight_list) / length(weight_list)  

```
# tf algo
```{r message=FALSE, warning=FALSE}
# Install if needed
BiocManager::install("dorothea")
BiocManager::install("viper")
library(dorothea)
library(viper)

# Load human TF-target interactions
data(dorothea_hs, package = "dorothea")

# Filter for high-confidence (A-C)
regulons <- dorothea_hs %>% dplyr::filter(confidence %in% c("A", "B", "C"))

# Extract normalized data from Seurat
expr_matrix <- as.matrix(GetAssayData(seurat_obj, slot = "data"))

# Run TF activity inference
tf_activities <- viper(expr_matrix, regulons, verbose = FALSE)

# Filter regulons for entries where the target gene is in your DEG list
deg_regulon_links <- regulons %>% dplyr::filter(target %in% deg_genes)

# Now extract the list of unique TFs that regulate at least one DEG
upstream_tfs <- unique(deg_regulon_links$tf)

# View a few
head(upstream_tfs)

```